claat/serve.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"archive/zip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+// cmdServe is the "claat serve ..." subcommand.
+func cmdServe() {
+	var depsDir = "bower_components"
+	var elemDir = "elements"
+	var elemFile = elemDir + "/codelab.html"
+	const codelabElem = `
+<link rel="import" href="../bower_components/google-codelab-elements/google-codelab-elements.html">
+`
+	err := os.MkdirAll(depsDir, 0755)
+	if err != nil {
+		fatalf(err.Error())
+	}
+	// Go get the dependencies.
+	err = fetchRepo(depsDir, "googlecodelabs/codelab-components#2.0.2")
+	if err != nil {
+		fatalf(err.Error())
+	}
+	os.Rename(depsDir+"/codelab-components", depsDir+"/google-codelab-elements")
+	os.Rename(depsDir+"/code-prettify", depsDir+"/google-prettify")
+	err = os.MkdirAll(elemDir, 0755)
+	if err != nil {
+		fatalf(err.Error())
+	}
+	if _, err := os.Stat(elemFile); os.IsNotExist(err) {
+		f, err := os.Create(elemFile)
+		if err != nil {
+			fatalf(err.Error())
+		}
+		defer f.Close()
+		f.WriteString(codelabElem)
+	}
+
+	http.Handle("/", http.FileServer(http.Dir(".")))
+	fmt.Println("Serving on " + *addr + ", opening browser tab now...")
+	openBrowser("http://" + *addr)
+	fatalf("claat serve: %+v", http.ListenAndServe(*addr, nil))
+}
+
+// downloadFile will download a url to a local file. It's efficient because it will
+// write as it downloads and not load the whole file into memory.
+func downloadFile(filepath string, url string) error {
+	fmt.Println("Downloading " + url)
+	out, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func fetchRepo(depsDir string, spec string) error {
+	if spec == "^0.7.2" {
+		spec = "webcomponents/webcomponentsjs#^0.7.2"
+	}
```

Comment:
What...?! :)
And if another transient dep had something like that in its bower.json?

I think a better way would be for `fetchRepo` to have a signature like this:

```go
// fetchRepo downloads a repo from github and unpacks it into basedir/dest,
// calling itself recursively for every dependency found in the unpacked bower.json.
// If the basedir/dest directly already exists, fetchRepo does nothing.
//
// The spec has the following format: "githuborg/repo#version".
// Any non-alphanumeric are stripped from the version.
// If version is missing, "master" is used.
// The bower registry is not used for naming resolution.
func fetchRepo(basedir, dest, spec string) error
```

This way you won't have to `os.Rename(depsDir+"/code-prettify", depsDir+"/google-prettify")` above, but instead just call:

    fetchRepo(depsDir, "google-codelab-elements", "googlecodelabs/codelab-components#2.0.2")

I would also add a map for exceptions:

```go
// bowerComp maps non-conforming spec as described in fetchRepo doc comments.
// It is keyed by the name found in a bower.json dependency list,
// with values corresponding to a valid githuborg/repo spec.
bowerComp = map[string][string]{
  "webcomponentsjs":  "webcomponents/webcomponentsjs"
}
```

Reply:
If I encode dest in the fetchRepo arg list, then I need to parse the spec to find the desired dest (in most cases). Wouldn't it be cleaner to pass just basedir and spec, and then have fetchRepo parse spec (which it already does anyway) to find the desired dest, and use the exception map to override the dest for cases where the dest needs to differ from the repo name? That way I don't have to parse the spec both inside and outside of fetchRepo.

Reply:
Actually, I realized that sometimes I want to look for exceptions based on comp name and sometimes based on spec. So I reverted to the simpler comp->spec map and I use it naturally for the one test and in a reverse fashion for the other case. It's inefficient to search a map for matching value but it's a small enough map (and very likely to stay small), I didn't think it was worth auto-building a reverse map.

Reply:
> If I encode dest in the fetchRepo arg list, then I need to parse the spec to find the desired dest (in most cases)

I think this should never happen, i.e. parsing spec to find the desired dest. Bower deps always have the "dest" dir. For instance, if you have this:

```json
{
  "dependencies": {
    "google-codelab-elements": "googlecodelabs/codelab-components#2.0.2"
  }
}
```
The dest dir is "google-codelab-elements". In general, it's always the key in map.